I added a test case.

before:

```sh
$ make functionaltest TEST_FILE=test/functional/lua/diagnostic_spec.lua TEST_FILTER="handles diagnostics without extmark_id"                   
mkdir -p ".deps"
/usr/bin/cmake -S /home/ras/ghq/github.com/neovim/neovim//cmake.deps -B ".deps" -G "Ninja" 
-- Found GNU Make at /usr/bin/gmake
-- CMAKE_BUILD_TYPE=Release
-- Configuring done (0.2s)
-- Generating done (0.0s)
-- Build files have been written to: /home/ras/ghq/github.com/neovim/neovim/.deps
mkdir -p build
touch "build/.ran-deps-cmake"
/usr/bin/cmake --build ".deps"
ninja: no work to do.
/usr/bin/cmake --build build --target functionaltest
[0/2] Re-checking globbed directories...
[2/3] cd /home/ras/ghq/github.com/neovim/neovim/build/test && /usr/bin/cmake -D TE...b.com/neovim/neovim -P /home/ras/ghq/github.com/neovim/neovim/cmake/RunTests.cmake
-------- Global test environment setup.
-------- Running tests from test/functional/lua/diagnostic_spec.lua
RUN      T1 vim.diagnostic show() and hide() handles diagnostics without extmark_id: 1.00 ms ERR
test/functional/lua/diagnostic_spec.lua:613: .../github.com/neovim/neovim/runtime/lua/vim/diagnostic.lua:663: Invalid 'id': Expected Lua number

stack traceback:
        test/functional/testnvim/exec_lua.lua:124: in function 'exec_lua'
        test/functional/lua/diagnostic_spec.lua:613: in function <test/functional/lua/diagnostic_spec.lua:612>

-------- 1 test from test/functional/lua/diagnostic_spec.lua (20.00 ms total)

-------- Global test environment teardown.
======== 1 test from 1 test file ran. (21.00 ms total)
PASSED   0 tests.
ERROR    1 error, listed below:
ERROR    test/functional/lua/diagnostic_spec.lua @ 612: vim.diagnostic show() and hide() handles diagnostics without extmark_id
test/functional/lua/diagnostic_spec.lua:613: .../github.com/neovim/neovim/runtime/lua/vim/diagnostic.lua:663: Invalid 'id': Expected Lua number

stack traceback:
        test/functional/testnvim/exec_lua.lua:124: in function 'exec_lua'
        test/functional/lua/diagnostic_spec.lua:613: in function <test/functional/lua/diagnostic_spec.lua:612>


 1 ERROR
------------------------------------------------------------------------------
$NVIM_LOG_FILE: /home/ras/.local/state/nvim/log
(last 10 lines)
DBG 2025-12-31T01:05:22.228 T1.84697.0 inbuf_poll:514: blocking... events=true
DBG 2025-12-31T01:05:22.229 T1.84697.0 receive_msgpack:208: ch 1: parsing 545 bytes from msgpack Stream: 0x55fc542340a0
DBG 2025-12-31T01:05:22.229 T1.84697.0 RPC: <- 1: [request]   id=2: nvim_exec_lua
DBG 2025-12-31T01:05:22.229 T1.84697.0 handle_request:345: RPC: scheduled nvim_exec_lua
DBG 2025-12-31T01:05:22.229 T1.84697.0 state_enter:97: input: K_EVENT
DBG 2025-12-31T01:05:22.229 T1.84697.0 handle_nvim_exec_lua:7470: RPC: ch 1: invoke nvim_exec_lua
DBG 2025-12-31T01:05:22.229 T1.84697.0 RPC: -> 1: [response]  id=2
DBG 2025-12-31T01:05:22.229 T1.84697.0 may_trigger_safestate:305: SafeState: Start triggering
DBG 2025-12-31T01:05:22.229 T1.84697.0 inbuf_poll:514: blocking... events=false
DBG 2025-12-31T01:05:22.229 T1.84697.0 inbuf_poll:514: blocking... events=true
------------------------------------------------------------------------------
E5113: Lua chunk: 
-- Tests exited non-zero: 1
CMake Error at /home/ras/ghq/github.com/neovim/neovim/cmake/RunTests.cmake:100 (message):
  functional tests failed with error: 1


FAILED: test/CMakeFiles/functionaltest /home/ras/ghq/github.com/neovim/neovim/build/test/CMakeFiles/functionaltest 
cd /home/ras/ghq/github.com/neovim/neovim/build/test && /usr/bin/cmake -D TEST_TYPE=functional -D BUILD_DIR=/home/ras/ghq/github.com/neovim/neovim/build -D CIRRUS_CI= -
D CI_BUILD=OFF -D DEPS_INSTALL_DIR=/home/ras/ghq/github.com/neovim/neovim/build/usr -D NVIM_PRG=/home/ras/ghq/github.com/neovim/neovim/build/bin/nvim -D TEST_DIR=/home/
ras/ghq/github.com/neovim/neovim/test -D WORKING_DIR=/home/ras/ghq/github.com/neovim/neovim -P /home/ras/ghq/github.com/neovim/neovim/cmake/RunTests.cmake
ninja: build stopped: subcommand failed.
make: *** [Makefile:151: functionaltest] Error 1
```

after:

```sh
$ make functionaltest TEST_FILE=test/functional/lua/diagnostic_spec.lua TEST_FILTER="handles diagnostics without extmark_id"
mkdir -p ".deps"
/usr/bin/cmake -S /home/ras/ghq/github.com/neovim/neovim//cmake.deps -B ".deps" -G "Ninja" 
-- Found GNU Make at /usr/bin/gmake
-- CMAKE_BUILD_TYPE=Release
-- Configuring done (0.2s)
-- Generating done (0.0s)
-- Build files have been written to: /home/ras/ghq/github.com/neovim/neovim/.deps
mkdir -p build
touch "build/.ran-deps-cmake"
/usr/bin/cmake --build ".deps"
ninja: no work to do.
/usr/bin/cmake --build build --target functionaltest
[0/2] Re-checking globbed directories...
[2/3] cd /home/ras/ghq/github.com/neovim/neovim/build/test && /usr/bin/cmake -D TE...b.com/neovim/neovim -P /home/ras/ghq/github.com/neovim/neovim/cmake/RunTests.cmake
-------- Global test environment setup.
-------- Running tests from test/functional/lua/diagnostic_spec.lua
RUN      T1 vim.diagnostic show() and hide() handles diagnostics without extmark_id: 2.00 ms OK
-------- 1 test from test/functional/lua/diagnostic_spec.lua (19.00 ms total)

-------- Global test environment teardown.
======== 1 test from 1 test file ran. (20.00 ms total)
PASSED   1 test.
```

--- a/runtime/lua/vim/diagnostic.lua
+++ b/runtime/lua/vim/diagnostic.lua
@@ -655,6 +655,10 @@ local sign_highlight_map = make_highlight_map('Sign')
 --- @return integer end_col
 --- @return boolean valid
 local function get_logical_pos(diagnostic)
+  if not diagnostic._extmark_id then
+    return diagnostic.lnum, diagnostic.col, diagnostic.end_lnum, diagnostic.end_col, true
+  end
+
   local ns = M.get_namespace(diagnostic.namespace)
   local extmark = api.nvim_buf_get_extmark_by_id(
     diagnostic.bufnr,

--- a/test/functional/lua/diagnostic_spec.lua
+++ b/test/functional/lua/diagnostic_spec.lua
@@ -608,6 +608,21 @@ describe('vim.diagnostic', function()
         vim.diagnostic.hide(_G.diagnostic_ns)
       end)
     end)
+
+    it('handles diagnostics without extmark_id', function()
+      exec_lua(function()
+        vim.diagnostic.config({ virtual_text = true })
+
+        vim.diagnostic.set(_G.diagnostic_ns, _G.diagnostic_bufnr, {
+          _G.make_error('Error message', 0, 0, 0, 5),
+        })
+
+        local diags = vim.diagnostic.get(_G.diagnostic_bufnr, { namespace = _G.diagnostic_ns })
+        diags[1]._extmark_id = nil
+
+        vim.diagnostic.show(_G.diagnostic_ns, _G.diagnostic_bufnr, diags)
+      end)
+    end)
   end)
 
   describe('enable() and disable()', function()

